### PR TITLE
SOAR-0009 - Typesafe multipart with streaming

### PR DESCRIPTION
**Pull request by czechboy0**
### Motivation

Ready for a review, see rendered at: https://github.com/czechboy0/swift-openapi-generator/blob/hd-soar-0009/Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0009.md

### Modifications

N/A

### Result

N/A

### Test Plan

N/A

### CODE REVIEW

**Comment by simonjbeaumont** [comment]
This is now in review and will run until Nov 15th. Forum link: https://forums.swift.org/t/proposal-soar-0009-type-safe-streaming-multipart-support/68331.

File: Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0009.md [hunk 1]
```diff
@@ -0,0 +1,933 @@
...
+#### Boundary customization
+
+When sending a multipart message, the sender needs to choose a boundary string (for example, `___MY_BOUNDARY_1234__`) that is used to separate the individual parts. The boundary string must not appear in any of the parts themselves.
+
+With this proposal, we introduce a protocol called `MultipartBoundaryGenerator` with a single method `func makeBoundary() -> String`, which returns the boundary string. The method is called once per multipart message, so it's encouraged for implementations of `MultipartBoundaryGenerator` to return a somewhat random output every time the `makeBoundary` method is called.
+
+The runtime library comes with two implementations, `ConstantMultipartBoundaryGenerator` and `RandomizedMultipartBoundaryGenerator`.
+
+`ConstantMultipartBoundaryGenerator` returns the same boundary every time and is useful for testing and in cases where stable output for stable inputs is desired, for example for caching.
+`RandomizedMultipartBoundaryGenerator` uses a constant prefix and appends a random suffix made out of `0-9` digits, returning a different output every time.
```
Comment (dnadoba) [inline Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0009.md:410]: I would suggest calling it RandomMultipartBoundaryGenerator

Similar to the existing [`RandomNumberGenerator`](https://developer.apple.com/documentation/swift/randomnumbergenerator) protocol and [`SystemRandomNumberGenerator`](https://developer.apple.com/documentation/swift/systemrandomnumbergenerator) struct.
```suggestion
`RandomMultipartBoundaryGenerator` uses a constant prefix and appends a random suffix made out of `0-9` digits, returning a different output every time.
```

Comment (czechboy0) [reply Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0009.md:410]: Good suggestion, thanks

**Comment by nkarczmarzyk** [comment]
+1 It would be an enrichment to have the multipart generator merged in a timely manner. Good work!

**Comment by czechboy0** [comment]
@nkarczmarzyk As Si said, this work is already ongoing. I'm glad you're looking forward to it, it's a large feature with substantial changes in the generator, so once we release 1.0.0-alpha.1 next week, it would be great if you could give it a try in your project - we want to catch any final issues before publishing 1.0.0 two weeks later.

@simonjbeaumont - can you update the review please? It already finished, so this and the forums post can be updated.

**Comment by nkarczmarzyk** [comment]
> @nkarczmarzyk As Si said, this work is already ongoing. I'm glad you're looking forward to it, it's a large feature with substantial changes in the generator, so once we release 1.0.0-alpha.1 next week, it would be great if you could give it a try in your project - we want to catch any final issues before publishing 1.0.0 two weeks later.
> 
> @simonjbeaumont - can you update the review please? It already finished, so this and the forums post can be updated.

I am looking forward to it. I also took a look on the feature (wip) PR https://github.com/apple/swift-openapi-generator/pull/366/ . Great work in there! Seems like a very complex topic. When it is ready for testing, I will give it a try and give feedback. Until then I am trying to implement a workaround.

**Comment by czechboy0** [comment]
> > @nkarczmarzyk As Si said, this work is already ongoing. I'm glad you're looking forward to it, it's a large feature with substantial changes in the generator, so once we release 1.0.0-alpha.1 next week, it would be great if you could give it a try in your project - we want to catch any final issues before publishing 1.0.0 two weeks later.
> > @simonjbeaumont - can you update the review please? It already finished, so this and the forums post can be updated.
> 
> I am looking forward to it. I also took a look on the feature (wip) PR #366 . Great work in there! Seems like a very complex topic. When it is ready for testing, I will give it a try and give feedback. Until then I am trying to implement a workaround.

I just got everything working end to end, so you can try the hd-multipart branch of the generator + hd-multipart-converter-methods branch of the runtime library. I'll work to clean up the code and start landing it today.

**Comment by nkarczmarzyk** [comment]
@czechboy0 I have found https://github.com/czechboy0/swift-openapi-generator/tree/hd-multipart for the generator. Could you please give me an hint where to find hd-multipart-converter-methods for the runtime?

**Comment by czechboy0** [comment]
> @czechboy0 I have found https://github.com/czechboy0/swift-openapi-generator/tree/hd-multipart for the generator. Could you please give me an hint where to find hd-multipart-converter-methods for the runtime?

Since I wrote that comment, we already landed it all on main - so you can use that branch.

**Comment by nkarczmarzyk** [comment]
I really don't want to blow up this pr but currently it seems like something is broken. Using latest of your forked generator (hd-multipart branch) and latest runtime I am getting following result.

open.yaml (v 3.0.1):
`
paths:
[...]
  /Post/addFile:
    post:
      tags:
        - Post
      requestBody:
        content:
          multipart/form-data:
            schema:
              required:
                - File
                - PostId
                - Type
              type: object
              properties:
                PostId:
                  type: string
                  format: uuid
                File:
                  type: string
                  format: binary
                Type:
                  $ref: '#/components/schemas/FileType'
            encoding:
              PostId:
                style: form
              File:
                style: form
              Type:
                style: form
      responses:
        '200':
          description: Success
[...]
components:
  schemas:
    FileType:
      enum:
        - Picture
        - Audio
      type: string
[...]
`

Generated type:
`
[...]
    /// - Remark: HTTP `POST /Post/addFile`.
    /// - Remark: Generated from `#/paths//Post/addFile/post`.
    public enum post_sol_Post_sol_addFile {
        public static let id: Swift.String = "post/Post/addFile"
        public struct Input: Sendable, Hashable {
            /// Creates a new `Input`.
            public init() {}
        }
[...]
`

It is also possible that I am doing something wrong. I am quite new to the world of swift.

**Comment by czechboy0** [comment]
@nkarczmarzyk Try to remove the `encoding` key, since none of the 3 parts are URL-encoded strings, this won't have an effect. If it still doesn't work then, I'll try to repro locally.

**Comment by czechboy0** [comment]
Also, the request bod must be marked as required, multipart bodies cannot be optional (which is the default).

**Comment by czechboy0** [comment]
One more question, is the generator emitting any diagnostics when running? That could hint at why it's skipping the content.

**Comment by czechboy0** [comment]
@nkarczmarzyk I verified in your example, that you just need to add `required: true` to the request body and then it works.

**Comment by nkarczmarzyk** [comment]
@czechboy0 Thanks for the advice. Setting `required: true` results to a correct generation. I will continue testing 👍

**Comment by nkarczmarzyk** [comment]
I just found out that you are not passing the correct parameter type in https://github.com/apple/swift-openapi-generator/blob/main/Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0009.md.

Instead of:
`let response = try await client.uploadPhoto(body: multipartBody)`

It should be:
`let response = try await client.uploadPhoto(body: .multipartForm(multipartBody))`

**Comment by nkarczmarzyk** [comment]
@czechboy0 
And actually I do not fully understand the aspect of type safety.  The generated types look like this:

```
[...]
                    public struct PostIdPayload: Sendable, Hashable {
                        public var body: OpenAPIRuntime.HTTPBody
                        /// Creates a new `PostIdPayload`.
                        ///
                        /// - Parameters:
                        ///   - body:
                        public init(body: OpenAPIRuntime.HTTPBody) {
                            self.body = body
                        }
                    }
                    case PostId(OpenAPIRuntime.MultipartPart<Operations.post_sol_Post_sol_addFile.Input.Body.multipartFormPayload.PostIdPayload>)
                    /// - Remark: Generated from `#/paths/Post/addFile/POST/requestBody/multipartForm/File`.
                    public struct FilePayload: Sendable, Hashable {
                        public var body: OpenAPIRuntime.HTTPBody
                        /// Creates a new `FilePayload`.
                        ///
                        /// - Parameters:
                        ///   - body:
                        public init(body: OpenAPIRuntime.HTTPBody) {
                            self.body = body
                        }
                    }
                    case File(OpenAPIRuntime.MultipartPart<Operations.post_sol_Post_sol_addFile.Input.Body.multipartFormPayload.FilePayload>)
                    /// - Remark: Generated from `#/paths/Post/addFile/POST/requestBody/multipartForm/Type`.
                    public struct _TypePayload: Sendable, Hashable {
                        public var body: OpenAPIRuntime.HTTPBody
                        /// Creates a new `_TypePayload`.
                        ///
                        /// - Parameters:
                        ///   - body:
                        public init(body: OpenAPIRuntime.HTTPBody) {
                            self.body = body
                        }
                    }
                    case _Type(OpenAPIRuntime.MultipartPart<Operations.post_sol_Post_sol_addFile.Input.Body.multipartFormPayload._TypePayload>)
[...]
```

Couldn't I just pass any HTTPBody as payload, which is independent of the OpenAPI definition type?

**Comment by czechboy0** [comment]
Streaming types, such as raw strings and raw bytes are represented by the HTTPBody type. If you instead had, for example, a JSON object as a multipart part, you'd get a fully typesafe Codable struct generated.

**Comment by czechboy0** [comment]
> Hey @czechboy0 hope you had a great weekend. I am off testing again. Currently its very complicated to initialise the components in the correct way but I will report if I got more results.
> 
> Currently my code looks like this:
> 
> ```
>         let body = HTTPBody()
> 
>         let postIdPayload: MultipartPart<Operations.post_sol_Post_sol_addFile.Input.Body.multipartFormPayload.PostIdPayload> = MultipartPart(payload: Operations.post_sol_Post_sol_addFile.Input.Body.multipartFormPayload.PostIdPayload(body: body))
>         
>         let filePayload: MultipartPart<Operations.post_sol_Post_sol_addFile.Input.Body.multipartFormPayload.FilePayload> = MultipartPart(payload: Operations.post_sol_Post_sol_addFile.Input.Body.multipartFormPayload.FilePayload(body: body))
>         
>         let typePayload: MultipartPart<Operations.post_sol_Post_sol_addFile.Input.Body.multipartFormPayload._TypePayload> = MultipartPart(payload: Operations.post_sol_Post_sol_addFile.Input.Body.multipartFormPayload._TypePayload(body: body))
> 
>         let multipartBody: OpenAPIRuntime.MultipartBody<Operations.post_sol_Post_sol_addFile.Input.Body.multipartFormPayload> = .init(
>             [.PostId(postIdPayload), .File(filePayload), ._Type(typePayload)])
>         
>         let response = try await ClientProvider.sharedClient.post_sol_Post_sol_addFile(body: .multipartForm(multipartBody))
> ```

Right, I'd probably write it like this instead:

```
let response = try await ClientProvider.sharedClient.post_sol_Post_sol_addFile(body: .multipartForm([
    .PostId(.init(payload: .init(body: "my-post-id"))),
    .File(.init(payload: .init(body: "<file contents go here>"))),
    ._Type(.init(payload: .init(body: "my-type"))),
]))
```

**Pull request merged**